### PR TITLE
storage.storageImageDestination.Commit(): leverage image options

### PR DESCRIPTION
**Pull request by nalind**
When committing an image, pass big data items and the image metadata string in as part of the set of options we pass to the storage library when creating the image.

Fall back to setting them individually, as we would have originally, only if we end up updating an image that was already in local storage.

### CODE REVIEW

File: storage/storage_dest.go
Comment (mtrmac) [inline storage/storage_dest.go:917]: This can also be done using a parameter to `CreateImage` if the image does not exist yet; that would avoid another trip through the fairly costly locking implementation.

---

… and then we would really need the `commitSucceeded` handling _only_ in the “image already exists” case, and that would be an opportunity revisit whether we really want to delete the image, or maybe we can just delete that code.

**Review by mtrmac — COMMENTED** [review]
Thanks, this is a great improvement, both in correctness and efficiency.

---

Compare https://github.com/containers/storage/issues/595 — from a quick check this seems to be a major improvement in that problem space, but maybe we should order the “big data” items in a specific order (config, individual manifest, top-level manifest; what about signatures?!) to avoid the remaining races.

**Comment by vrothberg** [comment]
Can we do test-vendor PRs in Buildah and Podman before merging?

File: storage/storage_dest.go
Comment (nalind) [reply storage/storage_dest.go:795]: Good point; we already handle the image's manifest, and the one for the list it's in if the source reference is for a list, so we know that everything else is either a config blob or something else.  Switching this to just calling `digest.Canonical.FromBytes()` directly.

File: storage/storage_dest.go
Comment (nalind) [reply storage/storage_dest.go:821]: Thanks, switching to referencing that instead.

File: storage/storage_dest.go
Comment (nalind) [reply storage/storage_dest.go:810]: No worries, removing it as a temporary variable in a number of places.

File: storage/storage_dest.go
Comment (nalind) [reply storage/storage_dest.go:897]: Adding a note here:
```
FIXME: this _replaces_ any "signatures" blobs and their sizes (tracked in the metadata) which might have already been present with new values, when ideally we'd find a way to merge them since they all apply to the same image
```

File: storage/storage_dest.go
Comment (nalind) [reply storage/storage_dest.go:899]: It's mainly logging it because we don't get the old "saved image metadata" message, which always logs it, any more if `CreateImage()` doesn't find an already-present image.  I could break them up into two `logrus.Debug()` calls if that would make it look less weird.

File: storage/storage_dest.go
Comment (nalind) [reply storage/storage_dest.go:917]: Some care has to be taken with that parameter - if any of the names passed at create-time are already in use, we get an `ErrDuplicateName` error back, and I think we want to avoid that.

File: storage/storage_dest.go
Comment (mtrmac) [reply storage/storage_dest.go]: Oh… IIRC I didn’t notice that including the full metadata here is pre-existing.

Preserving that behavior in this PR is certainly reasonable.

It might make sense to remove that, but _even if_ we decided to do that, doing that in a separate PR would be better anyway.

File: storage/storage_dest.go
Comment (mtrmac) [reply storage/storage_dest.go:909]: Oops, I didn’t notice that `CreateImage` returns `ErrDuplicateName`.

It might make sense to let callers opt in into overriding names with a new field in `storage.ImageOptions`, but that would not be quite an one-line change, so let’s not wait on that.

**Review by mtrmac — COMMENTED** [review]
LGTM. Thanks!

Valentin asked for test-vendor PRs, that seems to be a reasonable precaution to me.

**Comment by TomSweeneyRedHat** [comment]
LGTM once the Buildah/Podman tests run.

**Review by vrothberg — APPROVED** [review]
LGTM
https://github.com/containers/podman/pull/19525 is green

**Pull request merged**
